File: src/main/java/com/github/asufana/presentationjava/annotations/ViewAnnotation.java
package com.github.asufana.presentationjava.annotations;

import static java.util.Optional.*;
import static org.apache.commons.lang3.StringUtils.isNotEmpty;

import java.util.*;

public class ViewAnnotation {
    
    private final String name;
    private final String thClass;
    private final Integer thWidth;
    private final boolean thRight;
    private final String tdClass;
    private final Integer tdWidth;
    private final boolean tdRight;
    private final String valueMethod;
    
    public ViewAnnotation(final View annotation) {
        name = annotation.name();
        thClass = annotation.thClass();
        thWidth = annotation.thWidth();
        thRight = annotation.thRight();
        tdClass = annotation.tdClass();
        tdWidth = annotation.tdWidth();
        tdRight = annotation.tdRight();
        valueMethod = annotation.valueMethod();
    }
    
    public String name() {
        return name;
    }
    
    public Optional<String> valueMethod() {
        return isNotEmpty(valueMethod)
                ? of(valueMethod)
                : Optional.empty();
    }
    
    public String thTagString() {
        return String.format("<th%s%s%s>", thClassAttr(), thWidthAttr(), thStyleAttr());
    }
    
    protected String thClassAttr() {
        return isNotEmpty(thClass)
                ? String.format(" class=\"%s\"", thClass)
                : "";
    }
    
    protected String thWidthAttr() {
        return thWidth != 0
                ? String.format(" width=\"%s%s\"", thWidth, "%")
                : "";
    }
    
    protected String thStyleAttr() {
        return thRight
                ? " style=\"text-align:right\""
                : "";
    }
    
    public String tdTagString() {
        return String.format("<td%s%s%s>", tdClassAttr(), tdWidthAttr(), tdStyleAttr());
    }
    
    protected String tdClassAttr() {
        return isNotEmpty(tdClass)
                ? String.format(" class=\"%s\"", tdClass)
                : "";
    }
    
    protected String tdWidthAttr() {
        return tdWidth != 0
                ? String.format(" width=\"%s%s\"", tdWidth, "%")
                : "";
    }
    
    protected String tdStyleAttr() {
        return tdRight
                ? " style=\"text-align:right\""
                : "";
    }
    
}

File: src/main/java/com/github/asufana/presentation/annotations/ViewAnnotation.kt
package com.github.asufana.presentation.annotations

data class ViewAnnotation(private val annotation: View) {

    val name = annotation.name
    val valueMethod: String?
            = if (annotation.valueMethod.isNotEmpty()) annotation.valueMethod else null
    protected val thClass: String?
            = if (annotation.thClass.isNotEmpty()) annotation.thClass else null
    protected val thWidth = annotation.thWidth
    protected val thRight = annotation.thRight
    protected val tdClass: String?
            = if (annotation.tdClass.isNotEmpty()) annotation.tdClass else null
    protected val tdWidth = annotation.tdWidth
    protected val tdRight = annotation.tdRight


    fun thTagString(): String {
        return "<th${thClassAttr()}${thWidthAttr()}${thStyleAttr()}>"
    }

    protected fun thClassAttr(): String {
        return thClass?.let { " class=\"$thClass\"" } ?: ""
    }

    protected fun thWidthAttr(): String {
        return if (thWidth != 0) " width=\"$thWidth%\"" else ""
    }

    protected fun thStyleAttr(): String {
        return if (thRight) " style=\"text-align:right\"" else ""
    }

    fun tdTagString(): String {
        return "<td${tdClassAttr()}${tdWidthAttr()}${tdStyleAttr()}>"
    }

    protected fun tdClassAttr(): String {
        return tdClass?.let { " class=\"$tdClass\"" } ?: ""
    }

    protected fun tdWidthAttr(): String {
        return if (tdWidth != 0) " width=\"$tdWidth%\"" else ""
    }

    protected fun tdStyleAttr(): String {
        return if (tdRight) " style=\"text-align:right\"" else ""
    }

}

